### PR TITLE
[DependencyInjection] Handle void return types in closure-proxy

### PR DESCRIPTION
**Pull request by pierredup**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | N/A

I recently got an error when registering an event listener that specifies a `void` return type. Dumping the container generates a closure proxy that always returns a value, which then conflicts with the return type hint.

E.G the following code is generated (some class names removed for readability)

```
$instance->addListener('kernel.view', /** @closure-proxy ... */ function (...\GetResponseForControllerResultEvent $event): void {
    return ${($_ = isset($this->services[listener']) ? $this->services['listener'] : $this->get('listener')) && false ?: '_'}->onKernelView($event);
}, 128);
```

This then causes the error `A void function must not return a value in ...`

So void return types should be handled by removing the `return` inside the closure

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -1651,7 +1651,9 @@ private function dumpValue($value, $interpolate = true)
             }
             $signature = preg_replace('/^(&?)[^(]*/', '$1', InheritanceProxyHelper::getSignature($r, $call));
 
-            return sprintf("/** @closure-proxy %s::%s */ function %s {\n            return %s->%s;\n        }", $class, $method, $signature, $this->dumpValue($reference), $call);
+            $return = !method_exists($r, 'getReturnType') || 'void' !== (string) $r->getReturnType();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: we should call `InheritanceProxyHelper::getTypeHint()` here instead

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 2]
```diff
@@ -598,6 +598,21 @@ public function testClosureProxy()
     }
 
     /**
+     * @requires PHP 7.0
+     */
+    public function testClosureProxyWithVoidReturnType()
+    {
+        $container = include self::$fixturesPath.'/containers/container35.php';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]: please give this file to a meaningful name (we're changing policy here, numbering is conflict prone)

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 3]
```diff
@@ -598,6 +598,21 @@ public function testClosureProxy()
...
+
+        $container->compile();
+        $dumper = new PhpDumper($container);
+
+        $this->assertStringEqualsFile(self::$fixturesPath.'/php/services35.php', $dumper->dump());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]: ~~35~~

**Review by nicolas-grekas — COMMENTED** [review]
good catch

**Review by nicolas-grekas — APPROVED** [review]
:+1:

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 4]
```diff
@@ -598,6 +598,21 @@ public function testClosureProxy()
     }
 
     /**
+     * @requires PHP 7.0
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]: 7.1 actually

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/containers/container_dump_proxy_with_void_return_type.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace Symfony\Component\DependencyInjection\Tests\Fixtures\Container35;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/containers/container_dump_proxy_with_void_return_type.php]: 35 > Void ?

**Comment by nicolas-grekas** [comment]
Thank you @pierredup.

**Pull request closed without merging**